### PR TITLE
DOC: header='infer' is not working when there is no header, closes #17473

### LINKED ISSUE #17483: header='infer' is not working when there is no header
#### Code Sample

```python
import tempfile
import pandas as pd

df = pd.DataFrame({'AAA' : [4,5,6,7], 'BBB' : [10,20,30,40],'CCC' : [100,50,-30,-50]})
tmp = tempfile.mkstemp()[1]
df.to_csv(tmp, header=False)
df = pd.read_csv(tmp, header='infer')
print(df)
```

#### Output

```
0 | 4 | 10 | 100
--| --| -- | --
1 | 5 | 20 | 50
2 | 6 | 30 | -30
3 | 7 | 40 | -50


```

Pandas cannot infer that there is no header in csv files even though I use `header='infer'` argument in `read_csv`.

#### Expected Output

```

0 | 1 | 2  | 3
--| --| -- | --
0 | 4 | 10 | 100
1 | 5 | 20 | 50
2 | 6 | 30 | -30
3 | 7 | 40 | -50

```

I expect header inference to work properly and return a dataframe with 4 rows here. See `data.table` R package for a working header inference implementation[[1](https://github.com/Rdatatable/data.table/blob/master/src/fread.c#L638)].

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: 3.2.1
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.26
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 6.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by cmazzullo**
- [X] closes #17483

### CODE REVIEW

File: doc/source/io.rst [hunk 1]
```diff
@@ -103,15 +103,19 @@ Column and Index Locations and Names
...
-  specified will be skipped (e.g. 2 in this example is skipped). Note that
-  this parameter ignores commented lines and empty lines if
-  ``skip_blank_lines=True``, so header=0 denotes the first line of data
-  rather than the first line of the file.
+  Row number(s) to use as the column names, and the start of the
+  data. Default behavior is to infer the column names: if no names are
+  passed the behavior is identical to ``header=0`` and column names
+  are inferred from the first line of the file, if column names are
+  passed explicitly then the behavior is identical to
+  ``header=None``. Explicitly pass ``header=0`` to be able to replace
```
Comment (jreback) [inline doc/source/io.rst]: I would put a paragraph break here, its pretty long.

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -74,15 +74,19 @@
...
-    specified will be skipped (e.g. 2 in this example is skipped). Note that
-    this parameter ignores commented lines and empty lines if
-    ``skip_blank_lines=True``, so header=0 denotes the first line of data
-    rather than the first line of the file.
+    Row number(s) to use as the column names, and the start of the
+    data.  Default behavior is to infer the column names: if no names
+    are passed the behavior is identical to ``header=0`` and column
+    names are inferred from the first line of the file, if column
+    names are passed explicitly then the behavior is identical to
+    ``header=None``. Explicitly pass ``header=0`` to be able to
```
Comment (jreback) [inline pandas/io/parsers.py]: I would also add a note in Handling Column Names, maybe a ``::note`` box would be nice.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 2 above)
Comment (jreback) [reply pandas/io/parsers.py]: can you make this note the same as above?

**Comment by jreback** [comment]
@jorisvandenbossche if you'd review

File: doc/source/io.rst [hunk 3]
```diff
@@ -553,6 +558,14 @@ If the header is in a row other than the first, pass the row number to
     data = 'skip this skip it\na,b,c\n1,2,3\n4,5,6\n7,8,9'
     pd.read_csv(StringIO(data), header=1)
 
+.. note::
+
+  Default behavior is to infer the column names: if no names are
+  passed the behavior is identical to ``header=0`` and column names
+  are inferred from the first line of the file, if column names are
+  passed explicitly then the behavior is identical to
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: I would add: passed explicitly with the `header` keyword

File: doc/source/io.rst [hunk 4]
```diff
@@ -103,15 +103,20 @@ Column and Index Locations and Names
...
-  rather than the first line of the file.
+  Row number(s) to use as the column names, and the start of the
+  data. Default behavior is to infer the column names: if no names are
+  passed the behavior is identical to ``header=0`` and column names
+  are inferred from the first line of the file, if column names are
+  passed explicitly then the behavior is identical to
+  ``header=None``.
+
+  Explicitly pass ``header=0`` to be able to replace existing
+  names. The header can be a list of ints that specify row locations
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: I would make the paragraph break here at "The header can be ..." and leave the first sentence at the previou paragraph.

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! Added two small comments

**Comment by jreback** [comment]
can you rebase

**Comment by cmazzullo** [comment]
Sure thing @jreback

**Comment by jorisvandenbossche** [comment]
@cmazzullo Seems something went wrong with the rebase (many commits included on github now). In principle if you do exactly the following (assuming pandas-dev/pandas remote is called upstream):

```
git fetch upstream
git checkout gh-17483
git rebase upstream/master
git push origin gh-17483 --force
```

this should solve it.

**Comment by jorisvandenbossche** [comment]
@cmazzullo quickly did the rebase myself. Thanks for this clarification of the docs!

**Comment by cmazzullo** [comment]
Thanks!

**Pull request merged**